### PR TITLE
[Research] Automated testing

### PR DESCRIPTION
**Pull request by Soare-Robert-Daniel**
### Summary

This PR tries to implement automated testing (unit, e2e) to check for bugs and performance issues.

[WIP] Remarks:

- Unit testing works very well with simple functions or pure components that are WP-free.
- Testing block will require e2e for WP deps setup.
- Type addonotation with Typescript makes the testing much more straightforward. Usage of AI completion (Tabnine) speeds up the test writhing time.

### Screenshots 

----

### Test instructions

---- 

### Checklist before the final review

- [ ] Visual elements are not affected by independent changes.
- [ ] It is at least compatible with the [minimum WordPress version](https://wordpress.org/plugins/otter-blocks/).
- [ ] It loads additional script in frontend only if it is required.
- [ ] Does not impact the [Core Web Vitals](https://web.dev/vitals/).
- [ ] In case of deprecation, old blocks are safely migrated.
- [ ] It is usable in Widgets and FSE.

### CODE REVIEW

**Comment by Soare-Robert-Daniel** [comment]
Gutenberg test commands: https://github.com/WordPress/gutenberg/blob/ad4f2a2263ecf5849b25a45e2e92f276271aa1b6/package.json#L285-L302

e2e: https://cs.github.com/WordPress/gutenberg/blob/91d73e8f6dc041f5b8665434b110d4c165dc1697/packages/scripts/config/jest-e2e.config.js#L4

e2e Playwright: https://cs.github.com/WordPress/gutenberg/blob/91d73e8f6dc041f5b8665434b110d4c165dc1697/test/e2e/playwright.config.ts?q=test%2Fe2e%2Fplaywright.config.ts

Performance: https://cs.github.com/WordPress/gutenberg/blob/91d73e8f6dc041f5b8665434b110d4c165dc1697/packages/e2e-tests/jest.performance.config.js#L4

**Comment by Soare-Robert-Daniel** [comment]
⚠️ React Font Loader creates an npm deps conflict since it uses only react 15 while the rest of the packages use react 17, especially the one for testing.

**Comment by Soare-Robert-Daniel** [comment]
Testing Overview from WP: https://github.com/WordPress/gutenberg/blob/91d73e8f6dc041f5b8665434b110d4c165dc1697/docs/contributors/code/testing-overview.md#L261

**Comment by Soare-Robert-Daniel** [comment]
Inserting block e2e: https://cs.github.com/WordPress/gutenberg/blob/91d73e8f6dc041f5b8665434b110d4c165dc1697/packages/e2e-tests/specs/editor/various/inserting-blocks.test.js

**Comment by Soare-Robert-Daniel** [comment]
Gutenberg is migrating to Playwright and React Testing Library from Puppeteer and Enzyme.

Articles for testing knowledge

- https://www.smashingmagazine.com/2022/02/testing-pipeline-101-frontend-testing/
- https://www.smashingmagazine.com/2022/07/testable-frontend-architecture/
- https://github.com/WordPress/gutenberg/labels/Automated%20Testing
- https://www.lambdatest.com/blog/playwright-framework/
- https://make.wordpress.org/core/2019/06/27/introducing-the-wordpress-e2e-tests/#:~:text=The%20purpose%20of%20e2e%20(end,real%20user%20manipulating%20the%20interface.
- https://www.youtube.com/watch?v=dxbh0bl4It8&t=1905s

**Comment by selul** [comment]
we have also a cypress subscription if you want to consider

**Comment by Soare-Robert-Daniel** [comment]
Gutenberg is moving to Playwright activly. The current package for e2e is in maintenance mode.

Repo:  https://github.com/WordPress/gutenberg/tree/HEAD/packages/e2e-test-utils-playwright
Tracking the migration: https://github.com/WordPress/gutenberg/issues/38851

ℹ️  The package is unreleased to npm. We can not use it yet.

The exciting part: The Editor is given as a variable to the test function. This means that all the setup and cleaning are automatically done.

This is a migration example: https://github.com/WordPress/gutenberg/commit/3964d92fe33cbdabd98fee25221bc4a4c2da4871

File: src/blocks/test/e2e/performance/performance.spec.js [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+	// 		results.firstContentfulPaint.push( firstContentfulPaint );
+	// 		results.firstBlock.push( firstBlock );
+	// 	}
+	// });
+
+	it( 'Typing', async() => {
+
+		// Measuring typing performance.
+		await insertBlock( 'Paragraph' );
+		let i = 100;
```
Comment (Soare-Robert-Daniel) [inline src/blocks/test/e2e/performance/performance.spec.js:131]: @HardeepAsrani @selul with this values, the run will for aprox 100 * 2s = 200s. Can we go higher without creating problems for Github Actions limits?

Comment (HardeepAsrani) [reply src/blocks/test/e2e/performance/performance.spec.js:132]: Yea, it's fine. Github actions run for minutes, that's normal.

**Comment by HardeepAsrani** [comment]
> 0.97% of the values are lower than 60ms

@Soare-Robert-Daniel I'm curious if we can maybe have a drop-down list of things that are above 60ms? Also, is it 0.97 or 97%?

**Comment by Soare-Robert-Daniel** [comment]
> Also, is it 0.97 or 97%?

* 97%

We can have a list. But depends a lot on the testing session. Sometimes you get big spikes and influence the average. That is why I put the position of 60ms to see if we have some performance issues or if we have some outliners. 

One time a got two spikes for 500ms that got the average above 60 ms (then got the idea to add more info).

I will try to make a version with the list to see how it looks.

**Comment by selul** [comment]
@Soare-Robert-Daniel, maybe a p90 or p99 statistic could help filter the spikes?

**Comment by HardeepAsrani** [comment]
If you do add the list, make sure to use `<details><summary>Heading</summary>Hello world!</details>` so it doesn't take much space, like:

<details><summary>Heading</summary>Hello world!</details>

**Comment by Soare-Robert-Daniel** [comment]
@HardeepAsrani @selul do you think the Github machine influences many of the results? Those numbers look very good.

https://github.com/Codeinwp/otter-blocks/pull/1058#issuecomment-1210465167

**Comment by Soare-Robert-Daniel** [comment]
As a future node, the capabilities for e2e can be enhanced with Query Engine via offline support. We can use it to test migration from one version to another.

**Pull request merged**